kubelet: Don't ignore idsPerPod config

rata opened the pull request:
The idsPerPod where completely ignored since they were introduced in PR:
	https://github.com/kubernetes/kubernetes/pull/130028

The problem was the following:
	1. The userns manager (as well as all managers) is created before the config is copied to the kubelet object[1]
	2. The userns manager on creation is calling the kubelet_getter GetUserNamespacesIDsPerPod to get the idsPerPod
	3. The getter checks the configuration stored in the kubelet object, which hasn't been set at that point.
	4. As the config ss nil (unset), it returns the default value.

Therefore, the value was ignored.

To solve this, let's just pass the idsPerPod as a parameter to MakeUserNsManager(). This is the common pattern already used in the kubelet initialization.

[1]: https://github.com/kubernetes/kubernetes/blob/461ba83084ab7cb91ab692687bb7aedb05c6eb65/pkg/kubelet/kubelet.go#L1078-L1087
[2]: https://github.com/kubernetes/kubernetes/blob/461ba83084ab7cb91ab692687bb7aedb05c6eb65/pkg/kubelet/kubelet_getters.go#L145

cc @AkihiroSuda @giuseppe 

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

We now honor the idsPerPod config from the kubelet.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #133144 
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Fixes: #133144 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The kubelet now honors the configuration userNamespaces.idsPerPod. Before it was ignored.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

rata committed:
kubelet: Don't ignore idsPerPod config

The idsPerPod where completely ignored since they were introduced in PR:
	https://github.com/kubernetes/kubernetes/pull/130028

The problem was the following:
	1. The userns manager (as well as all managers) is created
	   before the config is copied to the kubelet object[1]
	2. The userns manager on creation is calling the kubelet_getter
	   GetUserNamespacesIDsPerPod to get the idsPerPod
	3. The getter checks the configuration stored in the kubelet
	   object, which hasn't been set at that point.
	4. As the config is nil (unset), it returns the default value.

Therefore, the value was ignored.

To solve this, let's just pass the idsPerPod as a parameter to
MakeUserNsManager(). This is the common pattern already used in the
kubelet initialization.

[1]: https://github.com/kubernetes/kubernetes/blob/461ba83084ab7cb91ab692687bb7aedb05c6eb65/pkg/kubelet/kubelet.go#L1078-L1087
[2]: https://github.com/kubernetes/kubernetes/blob/461ba83084ab7cb91ab692687bb7aedb05c6eb65/pkg/kubelet/kubelet_getters.go#L145

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

giuseppe reviewed (approved):
LGTM

SergeyKanzhelev commented:
Since this is not a GA feature, I do not think it is justifiable for 1.33 cherry-pick. However the KEP has an exception for 1.34 and the change is small enough to accept it. I am applying the milestone. If not merged by exception deadline, need to clear it up again.

/milestone v1.34

KEP: https://github.com/kubernetes/enhancements/issues/127

SergeyKanzhelev commented:
Is there any e2e tests can be written for this? There are many examples how e2e_node test can update kubelet config and test the outcome. Any chance you can implement it in this PR?

rata committed:
userns: Use better defaults if idsPerPod are used

The first UID used for userns mappings needs to be at least the
idsPerPod. When idsPerPod is extended to be 65536*2, for example, then
the default UID of 65536 doesn't work.

While this can be configured by the user, let's just improve the default
first UID to be the same as idsPerPod. This makes the default first UID
work out of the box if the user just wants to tune that.

This also simplifies testing, as we don't need to create a system user
and /etc/subuid and /etc/subgid files to test the idsPerPod setting.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
userns: Return nil manager if feature disabled

Golang allows to call methods on a nil object, as long as the methods
don't dereference the nil object. This is what we do here.

This makes all the userns configurations (idsPerPod or mapping configs
in /etc/subuid or /etc/subgid) to be ignored if the feature is off.

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata committed:
tests/e2e_node: Add test for userNamespaces.idsPerPod

Signed-off-by: Rodrigo Campos <rodrigoca@microsoft.com>

rata commented:
@SergeyKanzhelev Yeah, I doubted if a e2e test is worth it (now that it's a param is hard for it to not be present, and we have unit tests for the param). But fair, I've added and e2e tests now, it's the right call :).

Thanks for pointing out that it was easy to change the kubelet config with the e2e_node framework :-)

Please take another look :-)

rata commented:
/retest

ashutosh-engineer commented on pkg/kubelet/kubelet.go:
```diff
@@ -1075,12 +1075,17 @@ func NewMainKubelet(ctx context.Context,
 	})
 	klet.shutdownManager = shutdownManager
-	klet.usernsManager, err = userns.MakeUserNsManager(klet)
+	handlers = append(handlers, shutdownManager)
```
not looks good

rata replied:
It was moved above with all the shutdown manager stuff. Before this was in the middle without no good reason

AkihiroSuda commented on test/e2e_node/user_namespaces_test.go:
```diff
@@ -87,3 +100,69 @@ var _ = SIGDescribe("UserNamespaces", "[LinuxOnly]", feature.UserNamespacesSuppo
+	if err != nil {
+		return false
+	}
+	if string(outUids) != string(outGids) {
```
This case probably should raise an error?

AkihiroSuda commented on test/e2e_node/user_namespaces_test.go:
```diff
@@ -87,3 +100,69 @@ var _ = SIGDescribe("UserNamespaces", "[LinuxOnly]", feature.UserNamespacesSuppo
+	if string(outUids) != string(outGids) {
+		return false
+	}
+	return true
```
The number of the subids should be validated?

AkihiroSuda replied:
Notably, the stdout mustn't be empty

AkihiroSuda commented:
> @AkihiroSuda Are you back from PTO? Can you take the PR from here? :)

Yes, left some comments above

AkihiroSuda reviewed (approved):
LGTM, nits can be addressed later

jenshu commented:
As code freeze has passed, I'm removing this from the 1.34 milestone. Please let me know if you don't think it should be removed.

/milestone clear

rata commented:
Closing in favor of #133373
